File: server/src/lib/embeddings/ollama.ts
/**
 * Ollama Embedding Provider
 * Uses local Ollama server for embedding generation
 */

import type { EmbeddingProvider } from '../../types/index.js';

/** Default Ollama endpoint */
const DEFAULT_ENDPOINT = 'http://localhost:11434';

/** Default embedding model - nomic-embed-text has 768 dimensions */
const DEFAULT_MODEL = 'nomic-embed-text';

/** Model dimensions mapping */
const MODEL_DIMENSIONS: Record<string, number> = {
  'nomic-embed-text': 768,
  'mxbai-embed-large': 1024,
  'all-minilm': 384,
};

/** Model max chars (conservative estimates based on token limits) */
const MODEL_MAX_CHARS: Record<string, number> = {
  'nomic-embed-text': 5000,   // 2048 tokens * ~3.5 chars/token * 0.7 safety
  'mxbai-embed-large': 1500,  // 512 tokens
  'all-minilm': 800,          // 256 tokens
};

export class OllamaEmbeddingProvider implements EmbeddingProvider {
  readonly name = 'ollama';
  readonly model: string;
  readonly dimensions: number;
  readonly maxChars: number;
  private endpoint: string;

  constructor(options?: { endpoint?: string; model?: string }) {
    this.endpoint = options?.endpoint || process.env.OLLAMA_ENDPOINT || DEFAULT_ENDPOINT;
    this.model = options?.model || process.env.OLLAMA_MODEL || DEFAULT_MODEL;
    this.dimensions = MODEL_DIMENSIONS[this.model] || 768;
    this.maxChars = MODEL_MAX_CHARS[this.model] || 5000;
  }

  async isAvailable(): Promise<boolean> {
    try {
      const response = await fetch(`${this.endpoint}/api/tags`, {
        method: 'GET',
        signal: AbortSignal.timeout(2000),
      });

      if (!response.ok) return false;

      const data = await response.json() as { models?: Array<{ name: string }> };
      const models = data.models || [];

      // Check if our embedding model is available
      const hasModel = models.some(m =>
        m.name === this.model ||
        m.name.startsWith(`${this.model}:`)
      );

      return hasModel;
    } catch {
      return false;
    }
  }

  async generateEmbedding(text: string): Promise<number[]> {
    const response = await fetch(`${this.endpoint}/api/embed`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model: this.model,
        input: text,
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Ollama embedding failed: ${error}`);
    }

    const data = await response.json() as { embeddings: number[][] };

    if (!data.embeddings || data.embeddings.length === 0) {
      throw new Error('No embeddings returned from Ollama');
    }

    return data.embeddings[0];
  }

  async generateEmbeddings(texts: string[]): Promise<number[][]> {
    const response = await fetch(`${this.endpoint}/api/embed`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model: this.model,
        input: texts,
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Ollama batch embedding failed: ${error}`);
    }

    const data = await response.json() as { embeddings: number[][] };

    if (!data.embeddings) {
      throw new Error('No embeddings returned from Ollama');
    }

    return data.embeddings;
  }
}

File: cli/src/embeddings/ollama.rs
//! Ollama embedding provider.
//!
//! Uses local Ollama server for embedding generation.
//! This is the recommended provider for local development.

use crate::error::{Error, Result};
use serde::{Deserialize, Serialize};

use super::config::{resolve_ollama_endpoint, resolve_ollama_model};
use super::provider::EmbeddingProvider;
use super::types::{ollama_models, ProviderInfo};

/// Ollama embedding provider.
pub struct OllamaProvider {
    client: reqwest::Client,
    endpoint: String,
    model: String,
    dimensions: usize,
    max_chars: usize,
}

impl OllamaProvider {
    /// Create a new Ollama provider with default configuration.
    pub fn new() -> Self {
        Self::with_config(None, None)
    }

    /// Create a new Ollama provider with custom configuration.
    pub fn with_config(endpoint: Option<String>, model: Option<String>) -> Self {
        let endpoint = endpoint.unwrap_or_else(resolve_ollama_endpoint);
        let model = model.unwrap_or_else(resolve_ollama_model);
        let config = ollama_models::get_config(&model);

        Self {
            client: reqwest::Client::new(),
            endpoint,
            model,
            dimensions: config.dimensions,
            max_chars: config.max_chars,
        }
    }
}

impl Default for OllamaProvider {
    fn default() -> Self {
        Self::new()
    }
}

/// Ollama API response for listing models.
#[derive(Debug, Deserialize)]
struct OllamaTagsResponse {
    models: Option<Vec<OllamaModel>>,
}

#[derive(Debug, Deserialize)]
struct OllamaModel {
    name: String,
}

/// Ollama API request for embedding.
#[derive(Debug, Serialize)]
struct OllamaEmbedRequest<'a> {
    model: &'a str,
    input: EmbedInput<'a>,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum EmbedInput<'a> {
    Single(&'a str),
    Batch(Vec<&'a str>),
}

/// Ollama API response for embedding.
#[derive(Debug, Deserialize)]
struct OllamaEmbedResponse {
    embeddings: Vec<Vec<f32>>,
}

impl EmbeddingProvider for OllamaProvider {
    fn info(&self) -> ProviderInfo {
        ProviderInfo {
            name: "ollama".to_string(),
            model: self.model.clone(),
            dimensions: self.dimensions,
            max_chars: self.max_chars,
            available: false, // Will be checked by is_available()
        }
    }

    async fn is_available(&self) -> bool {
        let url = format!("{}/api/tags", self.endpoint);

        let response = match self.client
            .get(&url)
            .timeout(std::time::Duration::from_secs(2))
            .send()
            .await
        {
            Ok(r) => r,
            Err(_) => return false,
        };

        if !response.status().is_success() {
            return false;
        }

        let data: OllamaTagsResponse = match response.json().await {
            Ok(d) => d,
            Err(_) => return false,
        };

        // Check if our model is available
        data.models.map_or(false, |models| {
            models.iter().any(|m| {
                m.name == self.model || m.name.starts_with(&format!("{}:", self.model))
            })
        })
    }

    async fn generate_embedding(&self, text: &str) -> Result<Vec<f32>> {
        let url = format!("{}/api/embed", self.endpoint);

        let request = OllamaEmbedRequest {
            model: &self.model,
            input: EmbedInput::Single(text),
        };

        let response = self.client
            .post(&url)
            .json(&request)
            .send()
            .await
            .map_err(|e| Error::Embedding(format!("Ollama request failed: {e}")))?;

        if !response.status().is_success() {
            let error = response.text().await.unwrap_or_default();
            return Err(Error::Embedding(format!("Ollama embedding failed: {error}")));
        }

        let data: OllamaEmbedResponse = response.json().await
            .map_err(|e| Error::Embedding(format!("Failed to parse Ollama response: {e}")))?;

        data.embeddings.into_iter().next()
            .ok_or_else(|| Error::Embedding("No embeddings returned from Ollama".into()))
    }

    async fn generate_embeddings(&self, texts: &[&str]) -> Result<Vec<Vec<f32>>> {
        let url = format!("{}/api/embed", self.endpoint);

        let request = OllamaEmbedRequest {
            model: &self.model,
            input: EmbedInput::Batch(texts.to_vec()),
        };

        let response = self.client
            .post(&url)
            .json(&request)
            .send()
            .await
            .map_err(|e| Error::Embedding(format!("Ollama batch request failed: {e}")))?;

        if !response.status().is_success() {
            let error = response.text().await.unwrap_or_default();
            return Err(Error::Embedding(format!("Ollama batch embedding failed: {error}")));
        }

        let data: OllamaEmbedResponse = response.json().await
            .map_err(|e| Error::Embedding(format!("Failed to parse Ollama response: {e}")))?;

        Ok(data.embeddings)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ollama_provider_creation() {
        let provider = OllamaProvider::new();
        let info = provider.info();
        assert_eq!(info.name, "ollama");
        assert!(!info.model.is_empty());
        assert!(info.dimensions > 0);
    }

    #[test]
    fn test_ollama_provider_custom_config() {
        let provider = OllamaProvider::with_config(
            Some("http://custom:11434".to_string()),
            Some("mxbai-embed-large".to_string()),
        );
        let info = provider.info();
        assert_eq!(info.model, "mxbai-embed-large");
        assert_eq!(info.dimensions, 1024);
    }
}
